src/freshrssapi.cpp

```diff
@@ -0,0 +1,546 @@
+	if (reply == nullptr) {
+		LOG(Level::ERROR,
+			"FreshRssApi::get_subscribed_urls: failed to parse "
+			"response "
+			"as JSON.");
+		return urls;
+	}
+
+	json_object* subscription_obj{};
+	json_object_object_get_ex(reply, "subscriptions", &subscription_obj);
+	array_list* subscriptions = json_object_get_array(subscription_obj);
+
+	int len = array_list_length(subscriptions);
+
+	for (int i = 0; i < len; i++) {
+		std::vector<std::string> tags;
+		json_object* sub =
+			json_object_array_get_idx(subscription_obj, i);
+
+		json_object* id_str{};
+		json_object_object_get_ex(sub, "id", &id_str);
+		const char* id = json_object_get_string(id_str);
+
+		json_object* title_str{};
+		json_object_object_get_ex(sub, "title", &title_str);
+		const char* title = json_object_get_string(title_str);
+
+		tags.push_back(std::string("~") + title);
+
+		json_object* cats_obj{};
+		json_object_object_get_ex(sub, "categories", &cats_obj);
+		array_list* cats = json_object_get_array(cats_obj);
+
+		int ncats = array_list_length(cats);
+		for (int x = 0; x < ncats; x++) {
+			json_object* cat = json_object_array_get_idx(cats_obj, x);
+			json_object* cat_name{};
+			json_object_object_get_ex(cat, "label", &cat_name);
+			const char* category = json_object_get_string(cat_name);
+			tags.push_back(category);
+		}
+
+		char* escaped_id = curl_easy_escape(handle, id, 0);
+
+		auto url = strprintf::fmt("%s%s%s",
+				cfg->get_configvalue("freshrss-url"),
+				FRESHRSS_FEED_PREFIX,
+				escaped_id);
+		urls.push_back(TaggedFeedUrl(url, tags));
+
+		curl_free(escaped_id);
+	}
+
+	json_object_put(reply);
+
+	return urls;
+}
+
+void FreshRssApi::add_custom_headers(curl_slist** custom_headers)
+{
+	if (auth_header.empty()) {
+		auth_header = strprintf::fmt(
+				"Authorization: GoogleLogin auth=%s", auth);
+	}
+	LOG(Level::DEBUG,
+		"FreshRssApi::add_custom_headers header = %s",
+		auth_header);
+	*custom_headers =
+		curl_slist_append(*custom_headers, auth_header.c_str());
+}
+
+bool FreshRssApi::mark_all_read(const std::string& feedurl)
+{
+	std::string prefix =
+		cfg->get_configvalue("freshrss-url") + FRESHRSS_FEED_PREFIX;
+	std::string real_feedurl = feedurl.substr(
+			prefix.length(), feedurl.length() - prefix.length());
+	std::vector<std::string> elems = utils::tokenize(real_feedurl, "?");
+	try {
+		real_feedurl = utils::unescape_url(elems[0]);
+	} catch (const std::runtime_error& e) {
+		LOG(Level::DEBUG,
+			"FreshRssApi::mark_all_read: Failed to "
+			"unescape_url(%s): %s",
+			elems[0],
+			e.what());
+		return false;
+	}
+	std::string token = get_new_token();
+
+	std::string postcontent =
+		strprintf::fmt("s=%s&T=%s", real_feedurl, token);
+
+	std::string result = post_content(cfg->get_configvalue("freshrss-url") +
+			FRESHRSS_API_MARK_ALL_READ_URL,
+			postcontent);
+
+	return result == "OK";
+}
+
+bool FreshRssApi::mark_article_read(const std::string& guid, bool read)
+{
+	std::string token = get_new_token();
```

Comment:
Suggestion: keep a token in a class field, to avoid making a request each time you mark an article read. This might complicate the code though, as you'll need to handle token expiration.

Reply:
At present it seems that FreshRSS tokens [do not expire](https://github.com/FreshRSS/FreshRSS/blob/master/p/api/greader.php#L206). I'll experiment.

Reply:
I have implemented a solution, though when (or more likely, if) FreshRSS implements a proper token for its greader api it will start to break. It does improve speed however.